### ISSUE TITLE
How to handle the new par2cmdline version

### ISSUE THREAD

**Issue by Safihre** [sabnzbd/sabnzbd#902]
The new par2cmdline 0.7.0 is out. This one now also includes the OpenMP code, so it is also multi-threaded.

Now how are we going to incorporate this into SABnzbd? par2-tbb is still faster, so for the Windows/macOS releases it doesn't seem logical to completly switch to par2cmdline.
On Linux platforms we already detect if there is a multi-threaded par2 installed which we can use to our advantage.

So we could really make the `Multicore par2` switch force to only use 1 thread by adding the `-t` parameter to both par2cmdline (if multi-threaded is detected) and to par2-tbb.
In that case I would opt to only have a Special setting called `force_par2cmdline` and not offer this option to the user to not use par2-tbb on Windows/macOS?

**Comment by thezoggy** [comment]
par2-tbb on windows uses an old par2cmdline (0.4).. which means its behind the times in features and bugs that have been fixed still exist on it...  now it would awesome if there was a new par2-tbb build so we get the best of both worlds...  i know you ran tests before and found par2-tbb significantly faster than par2cmdline 0.6.13...  can you run them again against 0.7.0 ?
https://github.com/Parchive/par2cmdline/issues/27#issuecomment-170940850

**Comment by Safihre** [comment]
Good point. 
Test system: Core i5-4200 @ 2.30Ghz (2 cores, 4 threads), 12GB memory and all data on a fast SSD.

I used a dataset of 2GB with 42 missing articles throughout the download, what usually happens on usenet and ran the test twice for each program:

**par2cmdline 0.6.13 (only x86):**
1: `TotalSeconds      : 87.6838234`
2:  `TotalSeconds      : 86.2515236`

**par2cmdline 0.7.0 (x64):**
1: `TotalSeconds      : 61.612304`
2:  `TotalSeconds      : 60.4033289`

**par2tbb (x64):**
1: `TotalSeconds      : 49.0121549`
2:  `TotalSeconds      : 51.4033289`

**That's still pretty substantial.. But:**

### **Hypothesis:** par2cmdline 0.7.0 doesn't use multithreading to do verification, so the biggest difference is there.
Let's test verification only:

**par2cmdline 0.6.13 (only x86):**
1: `TotalSeconds      : 20.8311689`
2:  `TotalSeconds      : 20.9405503`

**par2cmdline 0.7.0 (x64):**
1: `TotalSeconds      : 19.0471533`
2:  `TotalSeconds      : 19.0547957`

**par2tbb (x64):**
1: `TotalSeconds      : 7.6788562`
2:  `TotalSeconds      : 7.5353623`

### There is a 13 second difference in the verification, which is more than the ~10 seconds difference observed in the full repair!

**Comment by Safihre** [comment]
Now for a Windows-only comparison I also tested repairing using Multipar.. **Wow**.

**multipar 1.2.9.9 (x64):**
1: `TotalSeconds      : 24.6406984`
2:  `TotalSeconds      : 25.5352605`

Just verification:
**multipar 1.2.9.9 (x64):**
1: `TotalSeconds      : 8.6350257`
2:  `TotalSeconds      : 8.7281912`

**Comment by thezoggy** [comment]
for par2cmdline 0.7.0 are you setting the number of threads `-t4` ?

**Comment by Safihre** [comment]
@thezoggy No need, it defaults to the number of available threads, I can also see it in the task manager that all cores are used during repair.
That's why it's 30% faster than the 0.6.13 version.

**Comment by thezoggy** [comment]
aye, repair is not verification..  well hopefully that gets fixed in 0.7.1. went to go submit GH issue for it but I see you already beat me to it

**Comment by Safihre** [comment]
So, with Multipar in place and surprisingly 0 bug reports so far, I would like to suggest to deprecate par2tbb in the bigger release that comes after 2.1.0 (e.g 2.2.0).
On Windows we would then just have Multipar and par2cmdline as fallback in case MultiPar is disabled. 
If there are really 0 problems with Multipar, we can drop both par2cmdline and par2tbb and only offer Multipar for Windows. 
So for now I removed the `Multicore par2` option from the Config > Switches page on Windows, it just looked confusing.

**Comment by Safihre** [comment]
@thezoggy I am leaning more and more to the idea of completely removing both par2cmdline and par2tbb in 2.2.0 for Windows and just use Multipar. There have been 0 problems reported with it. 
So for now keeping the current par2cmdline because it's still single threaded and activated by disabling "Multicore par2". Would be confusing otherwise.

**Comment by thezoggy** [comment]
since we dropped 32bit only cpu.. perhaps its time to start removing the 32bit tools and just have the 64bit versions (for things that dont have 64bit just keep the 32bit) that way we can cut down on included binaries.

also i personally think we should keep par2cmdline/par2tbb as failsafes for now until a few more releases since we dont have unittests and have to rely on user testing to know if its really safe to remove.. 

since we already have the code to support them and they dont take up that much space.. perhaps we should just offer up the ability for now to select which one you want sab to use.. so its easier to test/etc.. vs right now having to copy and replace binaries over each other.

**Comment by thezoggy** [comment]
once you upgrade par2cmdline it will be the multithreaded version..

**Comment by Safihre** [comment]
Yes exactly why I don't want to upgrade it. 
Regarding the 64bit drop, this is only for macOS because Apple itself dropped 32bit support. 

For Windows we will keep 32 and 64 of course!

**Comment by thezoggy** [comment]
you are wanting to remove tools because you feel they are duplicating functionality and dont offer any benefit over the other.

on the same vein, we should do that with the 32bit vs 64bit versions of the win utilities we include.
(which you already did for osx - minus the par2 which you keep a classic for fallback.)

if the x64 version of unrar has been proven to work well and there's no downsides to removing the 32bit version.. might as well save the 370kb? and simplify our code base.

32bit only cpu's are very very old.. and the os that support them are no longer supported by sab. 
the 64bit only cpus are very old and niche.. people that tried to run sab on them most likely would have already ran into issues because some of the included utilities only did 32bit (unzip). now those people most likely arent running windows.. so smaller chance of even being an issue.

https://en.wikipedia.org/wiki/X86#x86-64

**Comment by Safihre** [comment]
While normally indeed I'm all for removing things, in this case I have helped to many users in the past year that still run 32Bit Windows XP or some weird 32bit Windows server version. There were still 2000 downloads of the win32 release of 2.0.0, so somebody is using it 
In the installer the common files are not double, so not requiring extra space. We could indeed delete the 64bit tools from the win32 release.

**Comment by thezoggy** [comment]
interesting read and comments: https://www.backblaze.com/blog/64-bit-os-vs-32-bit-os/

**Comment by Safihre** [comment]
Agreed, but you'd really remove all 32bit stuff while it's so easy for us to create both? 
The numbers aren't high, but people are still using it: http://www.safihre.com/sabnzbd/stats.html

**Comment by Safihre** [comment]
In 2.1.0 the option to disable Multipar was removed on Windows and on macOS we removed the option to disable Multicore par2.
With 2.1.0 release [passing 100.000 downloads](http://www.safihre.com/sabnzbd/stats.html) I suggest the following changes will be made:

- [ ] Remove `par2-classic` for macOS
- [ ] Remove `par2cmdline` for Windows but keep `par2tbb` as fallback, `par2cmdline` cannot handle long-paths on Windows so it's really useless. It doesn't hurt to keep `par2tbb` as backup.
- [ ] Remove 64bit `par2tbb` for Windows to save space, since now it's only a fallback.
- [ ] Remove `par2_multicore` switch.
